pycheribuild/projects/run_qemu.py

```diff
@@ -413,28 +419,59 @@ def setup_config_options(cls, **kwargs):
+        cls.list_kernels = cls.add_bool_option("list-kernels", show_help=False,
+                                               help="List available kernel configs to run and exit")
+        cls.kernel_config = cls.add_config_option(
+            "kernel-config", show_help=True,
+            help="Select the kernel to run by specifying the kernel build configuration name."
+                 "The list of available kernel configurations is given by --list-kernels")
 
     def __init__(self, config: CheriConfig, source_class: "typing.Type[BuildFreeBSD]" = None,
                  disk_image_class: "typing.Type[BuildFreeBSDImage]" = None, needs_disk_image=True):
         super().__init__(config)
         if source_class is None and disk_image_class is not None:
             # noinspection PyProtectedMember
-            source_class = disk_image_class.get_instance(self).source_project
-        self.source_class = source_class
-        self.current_kernel = source_class.get_installed_kernel_path(self)
+            self.source_project = disk_image_class.get_instance(self).source_project
+        else:
+            self.source_project = source_class.get_instance(self)
+
+        if self.kernel_config:
+            if self.kernel_config not in self._valid_kernel_configs():
+                self.fatal("Selected kernel configuration", self.kernel_config, "is not available")
+                self._list_kernel_configs()
+        else:
+            self.kernel_config = self.source_project.default_kernel_config(ConfigPlatform.QEMU)
+        self.current_kernel = self.source_project.get_kernel_install_path(self.kernel_config)
+
+        self._project_specific_options += ["-append", "kern.module_path={}".format(self.current_kernel.parent)]
         if hasattr(source_class, "get_rootfs_dir"):
             # noinspection PyCallingNonCallable
             self.rootfs_path = source_class.get_rootfs_dir(self, config=config)
         if needs_disk_image:
             self.disk_image = disk_image_class.get_instance(self).disk_image_path
 
+    def _valid_kernel_configs(self):
+        return self.source_project.get_kernel_configs(platform=ConfigPlatform.QEMU)
+
+    def _list_kernel_configs(self):
+        self.info("Available kernels for qemu:")
+        for conf in self._valid_kernel_configs():
+            path = self.source_project.get_kernel_install_path(conf)
+            if conf == self.source_project.kernel_config:
+                self.info("*", conf, path)
+            else:
+                self.info(conf, path)
+
     def _copy_kernel_image_from_remote_host(self):
         scp_path = os.path.expandvars(self.remote_kernel_path)
         self.info("Copying kernel image from build machine:", scp_path)
         self.makedirs(self.current_kernel.parent)
         self.copy_remote_file(scp_path, self.current_kernel)
 
     def process(self):
+        if self.list_kernels:
+            self._list_kernel_configs()
+            sys.exit(1)
```

Comment:
1 is wrong, it ran successfully. But this is very inflexible; what happens if I want, say, `--run/list-kernels run-mips64-purecap run-riscv64-purecap`? IMO this should be a `return` not a `sys.exit`.